### ISSUE TITLE
[$2000] Login page content jumps when the connection status is changed

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#13876]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Goto https://staging.new.expensify.com/
2. Log out (if not), and now you should be on login page
3. Toggle the internet connection (On/off wifi)

## Expected Result:
There should be no content shift

## Actual Result:
There is a content shift

## Workaround:
unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?

- [ ] MacOS / Chrome / Safari

**Version Number:** 1.2.45-0
**Reproducible in staging?:** y
**Reproducible in production?:** y
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 

https://user-images.githubusercontent.com/43996225/209894031-14bc7f98-fcba-4cb9-8dca-17bfcea5925e.mov

https://user-images.githubusercontent.com/43996225/209894048-db733da0-62eb-4e6a-b323-356656fa88f0.mp4

**Expensify/Expensify Issue URL:**
**Issue reported by:** @mdneyazahmad 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1672221383592529

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01bad9df8c7ab2febd</li>
        <li>Upwork Job ID: 1608950026457661440</li>
        <li>Last Price Increase: 2023-01-06</li>
    </ul>
</details>

**Comment by MitchExpensify** [comment]
Triaged - I was unable to reproduce this on Chrome but could on Safari. Labeling external

**Comment by Prince-Mendiratta** [comment]
# Proposal
## Cause
When the network is online, the `OfflineIndicator` returns `null`. When the network goes offline, the offline indicator is inserted in the HTML, causing the other elements to jump up.

# Solution
## Solution 1
Instead of inserting the element at network change, let the element always exist but use the `visibility: true` css property by default to keep it hidden. Using `display: none` would still make the elements jump.

**Comment by allroundexperts** [comment]
### Proposal

## Issue
The following line is causing this issue:
https://github.com/Expensify/App/blob/6c81c5572908ba74e8072079b09a6f4c487ed329/src/pages/signin/SignInPageLayout/SignInPageContent.js#L58

## Fix
1.Add the above for small screen widths only.

**Comment by priyeshshah11** [comment]
# Proposal

## Problem
The View wrapping the `SignInPageForm` in `SignInPageContent.js` uses flexGrow2 which causes it to jump which is not necessary.

## Solution
Instead use `styles.flexGrow1` in that View.

https://github.com/Expensify/App/blob/d9ec05a2a26c42e865974322a2e9006b327f95d2/src/pages/signin/SignInPageLayout/SignInPageContent.js#L60

**Comment by jatinsonijs** [comment]
**Root Case:** 
This is happening due to different flex grow values of sibling. 
There is 2 sibling one with flex-grow: 1 and another one with flex-grow: 2

First empty view (with flex-grow: 1) we are using for add margin from top and also make it flexible for keyboard open and hide (it will shrink when keyboard open).

Another one main content with flex-grow: 2, so whenever there is new element added in it flex-grow: 2 will get more priority. in this case above with flex-grow: 1 view will shrink. We need to make both view with equal flex-grow value either 1 or 2.

```
{/* This empty view creates margin on the top of the sign in form which will shrink and grow depending on if the keyboard is open or not */}
<View style={[styles.flexGrow1, styles.signInPageContentTopSpacer]} />
```
```
<View style={[styles.flexGrow2]}>
                        <SignInPageForm style={[styles.alignSelfStretch]}>
 ```


**Proposal:**
```diff
diff --git a/src/pages/signin/SignInPageLayout/SignInPageContent.js b/src/pages/signin/SignInPageLayout/SignInPageContent.js
index 5f3ccb8a26..a85f499af9 100755
--- a/src/pages/signin/SignInPageLayout/SignInPageContent.js
+++ b/src/pages/signin/SignInPageLayout/SignInPageContent.js
@@ -57,7 +57,7 @@ const SignInPageContent = (props) => {
                     {/* This empty view creates margin on the top of the sign in form which will shrink and grow depending on if the keyboard is open or not */}
                     <View style={[styles.flexGrow1, styles.signInPageContentTopSpacer]} />
 
-                    <View style={[styles.flexGrow2]}>
+                    <View style={[styles.flexGrow1]}>
                         <SignInPageForm style={[styles.alignSelfStretch]}>
                             <View style={[
                                 styles.componentHeightLarge,
```

It is duplicate of @priyeshshah11 proposal but root case is missing. 

There is slightly difference in UI

Before:
![Simulator Screen Shot - iPhone 11 - 2022-12-31 at 08 57 41](https://user-images.githubusercontent.com/114683042/210124668-b51eb4af-f8c5-4eeb-a6b2-f8b31adf9292.png)

After:
 ![Simulator Screen Shot - iPhone 11 - 2022-12-31 at 08 57 25](https://user-images.githubusercontent.com/114683042/210124674-013f6c6f-c891-4af7-b562-1cb8c8db463f.png)

**Comment by MitchExpensify** [comment]
Not overdue, just waiting on proposal reviews from @rushatgabhane at this point

**Comment by umair-upwork** [comment]
### **Proposal**

**Problem**

This issue is happening due to `flexGrow2` style property in `<View>` wrapping  `<SignInPageForm />`. 
In `SignInPageContent.js` file under <KeyboardAvoidingView>, there are three `<View>` containers having different flex properties and some additional styling as well.

1. First `<View>` container has `flexGrow1` and `minHeight`, `maxHeight` as well.
2. Second `<View>` container has `flexGrow2`.
3. Third `<View>` container has `marginVertical`.

When connection status is changed, second `<View>` container jumps up a little due to `flexGrow2` will take up twice as much of the available positive free space as the flex item with `flexGrow1` when offline text appears at the bottom .

### **Solution**

We should not change anything in first and third `<View>` container because they have some additional styling such as margin and height. 

To fix this issue, we can add `flexAuto` property in second `<View>` instead of `flexGrow2` property because `flexAuto` can be used to make a flex item grow or shrink as necessary to fit the available space in the flex container, while also preserving its intrinsic size based on its content.

**Comment by jatinsonijs** [comment]
@umair-upwork I think it will do the same thing it will convert to flex-grow: 1

flex-grow: 1 mean:
flex-grow: 1
flex-shrink: 1
flex-basis: auto

flex: auto
flex: flex-grow flex-shrink flex-basis|auto|initial|inherit;
1 1 auto

ref:
https://www.w3schools.com/cssref/css3_pr_flex.php
https://stackoverflow.com/questions/35395691/understanding-the-difference-between-the-flex-and-flex-grow-properties

**Comment by umair-upwork** [comment]
@jatinsonijs   Yeah, it will do the same in some way.

`flex-grow: 1` allows the element to grow within the container if there is excess space, while `flex: auto` allows the element to grow or shrink within the container as needed.

`flex: auto` also sets the flex-shrink value to 1. This means that if the element needs to shrink in order to fit within the container, it will. However, it will still try to grow as much as possible within the container first.

And in first `<View>`, there is some height values attached to it in `styles.signInPageContentTopSpacer`. If you remove them and zoom out, you will see the content stays at the center which is not correct according `flex-grow `value.

So, i think it would be better to change only in second `<View>`.

**Comment by jatinsonijs** [comment]
no we will not remove any value we will just change flex-grow value either
both view flex-grow 1 or both view with flex-grow:2

**Comment by jatinsonijs** [comment]
we already have flexgrow 1 in first view

**Comment by umair-upwork** [comment]
@jatinsonijs  
`flex: auto` is a combination of `flex grow, basic, shrink`. So, either we use `flex grow 1` or `flex auto`, it will do the same thing and not change in first view. 

`flex grow` acts as single property where as `flex auto` acts  as combination of three.

We just need to change in second `<View>` with `flex auto` would be better approach i think so.

**Comment by wildan-m** [comment]
### Proposal

### Root Cause
This is expected flexbox behavior if we dynamically add contents in the middle of the flex item container.

I've created [a codepen](https://codepen.io/wildanmuhlis2/pen/bGjezOY) to reproduce it.

https://user-images.githubusercontent.com/11847279/210523033-01631746-356f-4bd9-8d98-4dcb1f81b922.mp4

This not only happened when the offline indicator was shown but also when any dynamic item is toggled (e.g. error message validation).

### Solution
To maintain the current layout position, we can add `flex-basis` to at least the first and second flex items.

The value of `flex-basis` will be the current percentage ratio from height resolution:
<img width="331" alt="image" src="https://user-images.githubusercontent.com/11847279/210524050-f3357d45-74f9-400d-aed0-f3ddba120b7e.png">

Since the last flex item has a 20px top margin we can reduce the second flex item basis by 20px.

### Required changes:
```diff
diff --git a/src/pages/signin/SignInPageLayout/SignInPageContent.js b/src/pages/signin/SignInPageLayout/SignInPageContent.js
index 5f3ccb8a2..ed6040498 100755
--- a/src/pages/signin/SignInPageLayout/SignInPageContent.js
+++ b/src/pages/signin/SignInPageLayout/SignInPageContent.js
@@ -57,7 +57,7 @@ const SignInPageContent = (props) => {
                     {/* This empty view creates margin on the top of the sign in form which will shrink and grow depending on if the keyboard is open or not */}
                     <View style={[styles.flexGrow1, styles.signInPageContentTopSpacer]} />
 
-                    <View style={[styles.flexGrow2]}>
+                    <View style={[styles.flexGrow2, styles.signInPageFormContainer]}>
                         <SignInPageForm style={[styles.alignSelfStretch]}>
                             <View style={[
                                 styles.componentHeightLarge,
diff --git a/src/styles/styles.js b/src/styles/styles.js
index b3de12803..e4fa8b062 100644
--- a/src/styles/styles.js
+++ b/src/styles/styles.js
@@ -923,6 +923,11 @@ const styles = {
     signInPageContentTopSpacer: {
         maxHeight: 132,
         minHeight: 24,
+        flexBasis: '11.10%',
+    },
+
+    signInPageFormContainer: {
+        flexBasis: 'calc(75.04% - 20px)',
     },
 
     signInPageLeftContainer: {
```

### Result

https://user-images.githubusercontent.com/11847279/210525696-451f6ab7-b884-4e5c-aad2-14845e81894d.mp4

**Comment by ali-thowfeek** [comment]
A little jump is not pleasant. But weirdly it kinda adds to the purpose of the alert by catching user's attention.
If we are to find the middle ground, we should smoothly animate the alert in, pushing the other content on top. This way it'll catch the user's attention and will look pleasant 😀

**Comment by MitchExpensify** [comment]
How do these proposals look to you @rushatgabhane ?

**Comment by wildan-m** [comment]
@rushatgabhane 

> The solution is a bit overkill though.

The calculation is one way to keep the current height proportion.

If we accept a little height adjustment, then we can change the flex-basis to a more round digit.

e.g.
flex container 1: 10%
flex container 2: calc(75%-20px)

**Comment by jatinsonijs** [comment]
> flex container 2: calc(75%-20px)

@wildan-m I it will not work for Android and iOS. App will crash. as calc style not exist for native.

**Comment by wildan-m** [comment]
@jatinsonijs nice catch!

we can change `calc` to a negative margin at the second flex item

```
    flex-basis: 75%;
    margin-bottom: -40px;
```

I've changed the -20px to -40px because we should also consider the 3rd flex item margin-bottom, otherwise we'll have extra empty space on the smallest screen.

complete changes:
https://github.com/wildan-m/App/tree/wildan-13876-login-page-content-jumps

**Comment by wildan-m** [comment]
@0xmiroslav the jumping behavior not only occurs when the offline indicator is shown, but it also occurs when field validation toggles. If we make the same `shouldKeepHeight ` props for the field validation message then we will see significant gaps between input elements

**Comment by jatinsonijs** [comment]
Agree with @wildan-m we cannot use same way for validation message. There is just little more space from top with flex-grow: 1 which not creating any regression. 
If we want to fix this issue, we can confirm from tgolen and design team about flex-grow: 2 - flex-grow: 1 as there is slightly more margin from top. attached screenshot here: https://github.com/Expensify/App/issues/13876#issuecomment-1368162343

**Comment by tgolen** [comment]
Hi all, I read through this issue and I wanted to make a few comments on things.

> @ali-thowfeek interesting idea. But I doubt animation is a priority right now.

Actually, animating the error/offline-indicator would be kinda cool and I know @shawnborton has been exploring how we can utilize more animation in the app. So, this might not be totally off the table.

In regards to using `flex:1` or `flex:2` for both containers, I do not feel strongly. Maybe it works OK, but I seem to remember there being a reason why I didn't go that route when I originally coded it. The problem escapes me now, but I want to throw out a word of caution as well:

**It's very easy to add regressions to other platforms with this layout, so whatever you propose, make sure it looks good on ALL PLATFORMS.** Many of these proposals I only see photos/videos for a single platform, but there are 6 platforms and they all have picky behavior.

The most important parts are:
1. The consistent margin at the top of the page
2. The top-margin shrinks when the keyboard is opened
3. The form button always displays above keyboard

**Comment by wildan-m** [comment]
@tgolen thanks for your feedback.

Any opinion about [my proposal](https://github.com/Expensify/App/issues/13876#issuecomment-1370691700)?

Do you think it will create critical regression?

I've tested it with various resolutions and made side-by-side comparisons. I think not much change in their layout. Please review it
cc: @rushatgabhane 

Please let me know if I should also test it with the native versions.

<details><summary>Desktop</summary>

Before|After
-|-
![image](https://user-images.githubusercontent.com/11847279/211451259-5aeae10b-6090-4d98-8d9e-5dda2c7bf6b7.png)|![image](https://user-images.githubusercontent.com/11847279/211451190-a9494e89-bf0a-44ad-8cd5-cd6d626e31a8.png)

</details>

<details><summary>Iphones</summary>

Before|After
-|-
<img width="1504" alt="Screenshot 2023-01-10 at 10 14 53" src="https://user-images.githubusercontent.com/11847279/211453981-51b25f75-01df-4932-8907-bdabd4629d4b.png">|<img width="1504" alt="Screenshot 2023-01-10 at 09 56 08" src="https://user-images.githubusercontent.com/11847279/211452308-aa8db794-01d8-4e02-a79b-1741a2e2f02f.png">


</details>

<details><summary>
Androids</summary>

Before|After
-|-
<img width="1504" alt="Screenshot 2023-01-10 at 10 14 41" src="https://user-images.githubusercontent.com/11847279/211454009-a9a68c9a-444d-4d4f-9d5c-49f2f76dab5f.png">|<img width="1504" alt="Screenshot 2023-01-10 at 09 55 51" src="https://user-images.githubusercontent.com/11847279/211452776-a1608ff6-886c-4b4f-8f25-e2aa3a4cbaf6.png">

</details>

<details><summary>Ipads</summary>

Before|After
-|-
<img width="1504" alt="Screenshot 2023-01-10 at 10 15 02" src="https://user-images.githubusercontent.com/11847279/211454065-6b816b96-ab49-4d64-a46b-137c337eea3d.png">|<img width="1504" alt="Screenshot 2023-01-10 at 09 56 20" src="https://user-images.githubusercontent.com/11847279/211453601-fbf3ab48-b3de-4dc5-9f63-cbcfc1d16ebb.png">


</details>

<details><summary>Tablets</summary>

Before|After
-|-
<img width="1504" alt="Screenshot 2023-01-10 at 10 15 24" src="https://user-images.githubusercontent.com/11847279/211454093-7f12ccab-a7c3-4284-9cfc-047676dc5382.png">|<img width="1504" alt="Screenshot 2023-01-10 at 09 56 33" src="https://user-images.githubusercontent.com/11847279/211453651-40492e1c-4d24-4643-a7a6-4372d3328249.png">


</details>

**Comment by jatinsonijs** [comment]
Here is the screenshot:

Desktop:
![desktop](https://user-images.githubusercontent.com/114683042/211463612-9c0e6195-5bef-4df4-92aa-0cb85397ecb7.png)

Desktop web:
![web](https://user-images.githubusercontent.com/114683042/211463664-a8666bf0-7bd3-43cf-a253-3ea870f94f48.png)

Android and iOS:
![android-ios](https://user-images.githubusercontent.com/114683042/211463699-cff6aadb-83ef-4c08-9312-1fb901c98589.png)

Android web:
![android-web](https://user-images.githubusercontent.com/114683042/211463759-cd4e9400-3274-44a1-a3c1-f081a34c0cad.png)

iOS web:
![ios-web](https://user-images.githubusercontent.com/114683042/211463795-b78225d7-edbb-45e0-85ff-76ca881c13f0.png)

Android web another size:
![android-web-2](https://user-images.githubusercontent.com/114683042/211463860-0b7802e1-6166-4382-9149-078bfb5d6dc9.png)

Android app very small screen:
![android](https://user-images.githubusercontent.com/114683042/211463913-ba601498-c2b2-49a0-9bfb-e2fd50fe1777.png)

**Comment by tgolen** [comment]
> Any opinion about https://github.com/Expensify/App/issues/13876#issuecomment-1370691700?

My initial feeling is that it feels overly complex. I'd prefer switching both containers to `flex:1` if that works. It's just a more simple solution that's easier to maintain.

I'm not sure about it creating regressions or not. I appreciate you testing and comparing on all platforms (I also love how you laid out the comparisons). Something that the comparison photos don't show is how the page reacts to the software keyboard, which was one of the critical parts I mentioned above.

**Comment by jatinsonijs** [comment]
@tgolen I have added screenshot with keyboard, can you plz look into this
https://github.com/Expensify/App/issues/13876#issuecomment-1376729213

**Comment by jatinsonijs** [comment]
Agree with @0xmiroslav, flex-grow will increase above view height, In web its not push content above in this case above view will not shrink user have to scroll to fully view Continue button. If we are not ok with this we should not do anything.

**Comment by MitchExpensify** [comment]
C+ and reporter invites sent! https://www.upwork.com/jobs/~01bad9df8c7ab2febd

**Comment by puneetlath** [comment]
@tylerkaraszewski @rushatgabhane @MitchExpensify this issue was created 2 weeks ago. Are we close to approving a proposal? If not, what's blocking us from getting this issue assigned? Don't hesitate to create a thread in #expensify-open-source to align faster in real time. Thanks!

**Comment by MitchExpensify** [comment]
I'll defer to @tylerkaraszewski and whether he agrees we can "do nothing" as suggested by @rushatgabhane & @0xmiroslav

**Comment by MitchExpensify** [comment]
Chatted with @tylerkaraszewski and we agree with doing nothing and closing this out
